src/layout/node.ts

```diff
@@ -17,29 +17,44 @@ export default function(node: Node, source: Source): void {
     }
 
     let call = dom.has(node) ? "update" : "add";
+    let parent = node.parentElement ? node.parentElement : null;
     switch (node.nodeType) {
         case Node.DOCUMENT_TYPE_NODE:
             let doctype = node as DocumentType;
             let docAttributes = { name: doctype.name, publicId: doctype.publicId, systemId: doctype.systemId };
             let docData = { tag: "*D", attributes: docAttributes };
-            dom[call](node, docData, source);
+            dom[call](node, parent, docData, source);
+            break;
+        case Node.DOCUMENT_FRAGMENT_NODE:
+            let shadowRoot = (node as ShadowRoot);
+            if (shadowRoot.host) {
+                // See: https://wicg.github.io/construct-stylesheets/
+                // For more details on adoptedStyleSheets for shadow dom
+                let style = "";
+                let adoptedStyleSheets: CSSStyleSheet[] = "adoptedStyleSheets" in shadowRoot ? shadowRoot["adoptedStyleSheets"] : [];
```

Comment:
>adoptedStyleSheets [](start = 59, length = 18)

This is only supported in recent Chromium browser ([https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot)). What happens to other ones Safari/Firefox/etc.?

Reply:
Shadow DOM is supported in many browsers (https://caniuse.com/#search=shadow) however adoptedStyleSheets (which is a way for web components to efficiently share CSS) is only supported in select browsers. Where it is not supported, they will be using classic style tag for styles and we already support that. Also for playback, we don't rely on adoptedStyleSheets since playback could be another browser and we turn these into regular style sheets (see render.ts). In short, we should be good cross browser.